src/firebase_functions/pubsub.py

```
@@ -32,9 +29,9 @@ class CloudEvent(Generic[T]):
 class Message(Generic[T]):
   message_id: str
   publish_time: dt.datetime
-  data: str
-  attributes: dict[str, str]
-  ordering_key: str
+  data: T
+  attributes: Optional[dict[str, str]] = None
+  ordering_key: Optional[str] = None
 
   @property
   def json(self) -> Dict[str, Any]:
```

This json implementation is incorrect, it should be attempting to JSON parse `self.data`: https://github.com/firebase/firebase-functions/blob/master/src/v2/providers/pubsub.ts#L102-L119

If json parse fails it should raise an error as in the JS impl there.

I'd suggest we support T for this if the user knows their JSON structure, e.g.:

`Message(T):`
`...`
`def json(self) -> T:`

---

Note though there is method on JS SDK called toJson: https://github.com/firebase/firebase-functions/blob/master/src/v2/providers/pubsub.ts#L121-L140 - which is doing what you currently doing here, but this is a method and not a property like https://github.com/firebase/firebase-functions/blob/master/src/v2/providers/pubsub.ts#L105-L119 

So I guess supporting both would be good, `self.json` for https://github.com/firebase/firebase-functions/blob/master/src/v2/providers/pubsub.ts#L105-L119 - and maybe `self.asdict()` for https://github.com/firebase/firebase-functions/blob/master/src/v2/providers/pubsub.ts#L126-L140